File: core/src/com/broll/gainea/server/core/map/Coordinates.java
package com.broll.gainea.server.core.map;

import java.awt.geom.Point2D;

public class Coordinates implements Comparable<Coordinates> {

    private float x;
    private float y;
    private float displayX;
    private float displayY;

    private float sx, sy;

    public Coordinates(float x, float y) {
        this.x = x;
        this.y = y;
    }

    public void shift(float x, float y) {
        this.sx += x;
        this.sy += y;
    }

    public void calcDisplayLocation(float size) {
        this.displayX = getX() * size;
        this.displayY = getY() * size;
    }

    public float getDisplayX() {
        return displayX;
    }

    public float getDisplayY() {
        return displayY;
    }

    public void setSx(float sx) {
        this.sx = sx;
    }

    public void setSy(float sy) {
        this.sy = sy;
    }

    public float getX() {
        return (sx + x);
    }

    public float getY() {
        return (sy + y);
    }

    public void mirrorY(float mirrorY) {
        this.y = mirrorY - y;
    }

    @Override
    public int compareTo(Coordinates o) {
        int compare = Float.compare(y, o.y);
        if (compare != 0) {
            return compare;
        }
        return Float.compare(x, o.x);
    }
}

File: server/src/main/com/broll/gainea/server/core/map/Coordinates.kt
package com.broll.gainea.server.core.map

class Coordinates(var rootX: Float, var rootY: Float) : Comparable<Coordinates> {
    var displayX = 0f
        private set
    var displayY = 0f
        private set
    private var sx = 0f
    private var sy = 0f

    var x: Float = 0f
        get() = sx + rootX
        private set

    var y: Float = 0f
        get() = sy + rootY
        private set

    fun shift(x: Float, y: Float) {
        sx += x
        sy += y
    }

    fun set(x: Float, y: Float) {
        this.rootX = x
        this.rootY = y
    }

    fun calcDisplayLocation(size: Float) {
        displayX = x * size
        displayY = y * size
    }

    fun setSx(sx: Float) {
        this.sx = sx
    }

    fun setSy(sy: Float) {
        this.sy = sy
    }

    fun mirrorY(mirrorY: Float) {
        rootY = mirrorY - rootY
    }

    override fun compareTo(o: Coordinates): Int {
        val compare = rootY.compareTo(o.rootY)
        return if (compare != 0) {
            compare
        } else rootX.compareTo(o.rootX)
    }
}
